uplink/src/base/actions/ota.rs

```diff
@@ -13,97 +57,179 @@ use crate::base::{Config, Stream};
+            Self {
+                config,
+                client,
+                ota_rx,
+                status_bucket,
+                bridge_tx,
+                sequence: 0,
+                action_id: String::default(),
+            },
+        ))
+    }
+
+    /// Spawn a thread to handle downloading OTA updates as per "update_firmware" actions and for
+    /// forwarding updated actions to bridge for further processing, i.e. update installation.
+    #[tokio::main(flavor = "current_thread")]
+    pub async fn start(mut self) -> Result<(), Error> {
+        loop {
+            self.sequence = 0;
+            // `_lock` is held until current action completes, hence failing all sends till it's dropped
+            let action = self.ota_rx.recv()?;
+            self.action_id = action.action_id.clone();
+
+            match self.run(action).await {
+                Ok(_) => {
+                    let status = ActionResponse::progress(&self.action_id, "Downloaded", 100)
+                        .set_sequence(self.sequence());
+                    self.send_status(status).await;
+                }
+                Err(e) => {
+                    let status = ActionResponse::failure(&self.action_id, e.to_string())
+                        .set_sequence(self.sequence());
+                    self.send_status(status).await;
+                }
+            }
+            // _lock is dropped here so that next OTA Action can be received
+        }
     }
 
-    async fn run(
-        &mut self,
-        ota_path: &PathBuf,
-        client: Client,
-        action: Action,
-        url: String,
-    ) -> Result<(), Error> {
+    // Accepts an OTA `Action` and performs necessary data extraction to actually download the OTA update
+    async fn run(&mut self, action: Action) -> Result<(), Error> {
         // Update action status for process initiated
         let status = ActionResponse::progress(&self.action_id, "Downloading", 0)
             .set_sequence(self.sequence());
         self.send_status(status).await;
 
-        // Create file to download files into
-        let file = self.create_file(ota_path).await?;
+        // Extract url and add ota_path in payload before recreating action to be sent to bridge
+        let Action { action_id, kind, name, payload } = action;
+        let mut update = serde_json::from_str::<FirmwareUpdate>(&payload)?;
+        let url = update.url.clone();
 
-        // Create handler to perform download from URL
-        let resp = client.get(url).send().await?;
+        // Ensure that directory for downloading file into, of the format `path/to/{version}/`, exists
+        let mut ota_path = PathBuf::from(self.config.ota.path.clone());
+        ota_path.push(&update.version);
+        create_dir_all(&ota_path)?;
+
+        // Create file to actually download into
+        let mut file_path = ota_path.to_owned();
+        let file_name = url.split("/").last().ok_or(Error::FileNameMissing(url.to_owned()))?;
```

Comment:
Is error handling necessary here given we are creating the directory and erroring out if we can't. When does this call return `None`?

Reply:
Ah we are creating the same file name as image name in url path

Reply:
Why not extract all of file creation into a method like the original branch?